pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,325 @@
+	// First check if we can add the command.
+	if err := q.CanAdd(providerIDs...); err != nil {
+		return fmt.Errorf("adding command, %w", err)
+	}
+	hash, err := Hash(cmd)
+	if err != nil {
+		return err
+	}
+	for _, candidate := range cmd.Candidates {
+		q.candidateProviderIDToCommandID[candidate.ProviderID()] = hash
+		q.idToCommand[hash] = cmd
+	}
+	// Idempotently mark for deletion
+	q.cluster.MarkForDeletion(providerIDs...)
+	q.RateLimitingInterface.Add(hash)
+	return nil
+}
+
+// Remove fully clears the queue of all references of a hash/command
+func (q *Queue) Remove(hash uint64, cmd *Command) {
+	q.RateLimitingInterface.Forget(hash)
+	for _, candidate := range cmd.Candidates {
+		delete(q.candidateProviderIDToCommandID, candidate.ProviderID())
+	}
+	delete(q.idToCommand, hash)
+	q.RateLimitingInterface.Done(hash)
+}
+
+// CanAdd is a quick check to see if the candidate is already part of a deprovisioning action
+func (q *Queue) CanAdd(ids ...string) error {
+	var err error
+	for _, id := range ids {
+		if _, ok := q.candidateProviderIDToCommandID[id]; ok {
+			err = multierr.Append(err, fmt.Errorf("candidate is part of active deprovisioning decision"))
+		}
+	}
+	return err
+}
+
+// Handle will check for timeouts, then execute a wait or termination.
+// If there
+func (q *Queue) Handle(ctx context.Context, cmd *Command) (bool, error) {
+	// logging.FromContext(ctx).Infof("DEBUGGING: This is the current time: %s, this is the time added: %s", q.clock)
+	if q.clock.Since(cmd.TimeAdded) > maxRetryDuration {
+		return false, fmt.Errorf("abandoning command, reached timeout, %w", cmd.LastError)
+	}
+	// If the time hasn't expired, either wait, or terminate.
+	requeue, err := q.WaitOrTerminate(ctx, cmd)
+	if err != nil {
+		// If there was an error, set this as the command's last error so that we can propagate it.
+		cmd.LastError = err
+		if !retry.IsRecoverable(err) {
+			return false, nil
+		}
+		return requeue, nil
+	}
+	return requeue, nil
+}
+
+// WaitOrTerminate will wait until launched machines are ready.
+// Once the replacements are ready, it will terminate the candidates.
+// Will return true if the item in the queue should be re-queued. If a command has
+// timed out, this will return false.
+// nolint:gocyclo
+func (q *Queue) WaitOrTerminate(ctx context.Context, cmd *Command) (bool, error) {
+	waitErrs := make([]error, len(cmd.ReplacementKeys))
+	for i := range cmd.ReplacementKeys {
+		key := cmd.ReplacementKeys[i]
+		// If we know the node claim is initialized, no need to check again.
+		if key.Initialized {
+			continue
+		}
+		// Get the nodeclaim
+		nodeClaim, err := nodeclaim.Get(ctx, q.kubeClient, key.Key)
+		if err != nil {
+			// The NodeClaim got deleted after an initial eventual consistency delay
+			// This means that there was an ICE error or the Node initializationTTL expired
+			// In this case, the error is unrecoverable, so don't requeue.
+			if errors.IsNotFound(err) && q.clock.Since(cmd.TimeAdded) > time.Second*5 {
```

Comment:
Does it still make sense to have this here or should we just give a little bit of time on the initial Add so we don't hit a case here where we don't read our own writes. That way, we wouldn't have to code this into the check directly